### PR TITLE
Crab netting ENG-788

### PR DESCRIPTION
**Pull request by sanandnarayan**
# Task:
Nets the queued deposits and withdrawals
New feature

### CODE REVIEW

File: packages/crab-netting/test/Netting.t.sol [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+        netting.depositUSDC(80 * 1e18);
+        assertEq(netting.usd_balance(depositor), 200e18);
+        vm.stopPrank();
+
+        vm.startPrank(withdrawer);
+        crab.approve(address(netting), 200 * 1e18);
+        netting.depositCrab(5 * 1e18);
+        netting.depositCrab(4 * 1e18);
+        netting.depositCrab(11 * 1e18);
+        assertEq(netting.crab_balance(withdrawer), 20e18);
```
Comment (haythem96) [inline packages/crab-netting/test/Netting.t.sol:57]: Those deposit calls should be done inside the test function for the related case, e.g inside `testNetting()` and `testNettingWithMultipleDeposits()`

Comment (haythem96) [reply packages/crab-netting/test/Netting.t.sol:57]: Also I think worth adding a simple test for the actual `depositCrab()` and `depositUSDC()`

File: packages/crab-netting/src/CrabNetting.sol [hunk 2]
```diff
@@ -7,49 +9,126 @@ contract CrabNetting {
...
+        // TODO make sure we dont need SafeMath in final version
+        uint256 i = depositsIndex;
+        while (_quantity > 0 && i < deposits.length) {
+            Receipt memory deposit = deposits[i];
+            if (deposit.amount < _quantity) {
+                _quantity = _quantity - deposit.amount;
+                usd_balance[deposit.sender] =
+                    usd_balance[deposit.sender] -
+                    deposit.amount;
+                IERC20(crab).transfer(deposit.sender, deposit.amount / _price);
```
Comment (haythem96) [inline packages/crab-netting/src/CrabNetting.sol:73]: Shouldn't this take into account decimals ?

File: packages/crab-netting/src/CrabNetting.sol [hunk 3]
```diff
@@ -7,49 +9,126 @@ contract CrabNetting {
...
+                i++;
+            } else {
+                deposits[i].amount = deposit.amount - _quantity;
+                usd_balance[deposit.sender] =
+                    usd_balance[deposit.sender] -
+                    _quantity;
+                IERC20(crab).transfer(deposit.sender, _quantity / _price);
```
Comment (haythem96) [inline packages/crab-netting/src/CrabNetting.sol:80]: Also here ?

File: packages/crab-netting/src/CrabNetting.sol [hunk 4]
```diff
@@ -7,49 +9,126 @@ contract CrabNetting {
...
+        IERC20(crab).transfer(msg.sender, _amount);
     }
 
-    function balanceOf(address account) public view returns (uint256) {
-        return usd_balance[account];
+    function netAtPrice(uint256 _price, uint256 _quantity) public {
+        // TODO make only owner
+        uint256 forWithdraw = _quantity;
+        require(
+            _quantity <= IERC20(usdc).balanceOf(address(this)),
```
Comment (haythem96) [inline packages/crab-netting/src/CrabNetting.sol:58]: We should make sure this one is 1e6, as in ur test u passing number with 18 decimals

File: packages/crab-netting/src/CrabNetting.sol [hunk 5]
```diff
@@ -7,49 +9,126 @@ contract CrabNetting {
...
+                    crab_balance[withdraw.sender] -
+                    withdraw.amount;
+                IERC20(usdc).transfer(
+                    withdraw.sender,
+                    withdraw.amount * _price
+                );
+                j++;
+            } else {
+                withdraws[j].amount = withdraw.amount - crab_quantity;
+                IERC20(usdc).transfer(withdraw.sender, crab_quantity * _price);
```
Comment (haythem96) [inline packages/crab-netting/src/CrabNetting.sol:102]: How are we handling decimals here ?

File: packages/crab-netting/src/CrabNetting.sol [hunk 6]
```diff
@@ -7,49 +9,126 @@ contract CrabNetting {
...
 
-    function withdrawUSDC(uint256 amount) public {
-        require(usd_balance[msg.sender] >= amount);
-        usd_balance[msg.sender] = usd_balance[msg.sender] - amount;
-        IERC20(usdc).transfer(msg.sender, amount);
+    function withdrawUSDC(uint256 _amount) public {
+        // TODO ensure final version does not need this check
+        //require(usd_balance[msg.sender] >= _amount, "Withdrawing more than balance");
+        usd_balance[msg.sender] = usd_balance[msg.sender] - _amount;
+        IERC20(usdc).transfer(msg.sender, _amount);
```
Comment (KMKoushik) [inline packages/crab-netting/src/CrabNetting.sol:38]: We are not updating the receipt here?

File: packages/crab-netting/src/CrabNetting.sol [hunk 7]
```diff
@@ -7,49 +9,126 @@ contract CrabNetting {
...
+        // TODO DO we need to do the same check for crab withdrawals?
+        // can we just do deposits alone and forget about withdrawals netting?
+
+        // TODO make sure we dont need SafeMath in final version
+        uint256 i = depositsIndex;
+        while (_quantity > 0 && i < deposits.length) {
+            Receipt memory deposit = deposits[i];
+            if (deposit.amount < _quantity) {
+                _quantity = _quantity - deposit.amount;
+                usd_balance[deposit.sender] =
```
Comment (KMKoushik) [inline packages/crab-netting/src/CrabNetting.sol:70]: Consider this situation
- I deposit 100 USDC, now the receipt will have 100 as amount and usdc_balance will be 100 
- I withdraw 50 USDC, now the receipt will still be 100 but usdc_balance will be 50
- this operation will lead to 50 - 100 and will throw error

File: packages/crab-netting/src/CrabNetting.sol (hunk 6 above)
Comment (KMKoushik) [inline packages/crab-netting/src/CrabNetting.sol:38]: We are not updating the receipt here?

File: packages/crab-netting/src/CrabNetting.sol [hunk 8]
```diff
@@ -7,49 +9,126 @@ contract CrabNetting {
...
+                    usd_balance[deposit.sender] -
+                    deposit.amount;
+                IERC20(crab).transfer(deposit.sender, deposit.amount / _price);
+                i++;
+            } else {
+                deposits[i].amount = deposit.amount - _quantity;
+                usd_balance[deposit.sender] =
```
Comment (KMKoushik) [inline packages/crab-netting/src/CrabNetting.sol:77]: Same comment as above

File: packages/crab-netting/src/CrabNetting.sol [hunk 9]
```diff
@@ -7,49 +9,126 @@ contract CrabNetting {
...
+            }
+        }
+        depositsIndex = depositsIndex + i;
+        uint256 j = withdrawsIndex;
+        uint256 crab_quantity = forWithdraw / _price;
+        while (crab_quantity > 0 && j < withdraws.length) {
+            Receipt memory withdraw = withdraws[j];
+            if (withdraw.amount < crab_quantity) {
+                crab_quantity = crab_quantity - withdraw.amount;
+                crab_balance[withdraw.sender] =
```
Comment (KMKoushik) [inline packages/crab-netting/src/CrabNetting.sol:92]: Same comment here

File: packages/crab-netting/src/CrabNetting.sol [hunk 10]
```diff
@@ -7,49 +9,126 @@ contract CrabNetting {
...
+                    withdraw.amount;
+                IERC20(usdc).transfer(
+                    withdraw.sender,
+                    withdraw.amount * _price
+                );
+                j++;
+            } else {
+                withdraws[j].amount = withdraw.amount - crab_quantity;
+                IERC20(usdc).transfer(withdraw.sender, crab_quantity * _price);
+                crab_balance[withdraw.sender] =
```
Comment (KMKoushik) [inline packages/crab-netting/src/CrabNetting.sol:103]: Same comment here

File: packages/crab-netting/src/CrabNetting.sol [hunk 11]
```diff
@@ -7,49 +9,126 @@ contract CrabNetting {
...
+    mapping(address => uint256) public crab_balance;
     struct Receipt {
-        address depositor;
+        address sender;
         uint256 amount;
     }
-    Receipt[] deposits;
-    Receipt[] crab_deposits;
+    Receipt[] public deposits;
+    uint256 public depositsIndex;
```
Comment (KMKoushik) [inline packages/crab-netting/src/CrabNetting.sol:19]: I see we are deleting the receipt element in netting, won't it make the index always 0?

File: packages/crab-netting/src/CrabNetting.sol [hunk 12]
```diff
@@ -7,49 +9,126 @@ contract CrabNetting {
...
+    mapping(address => uint256) public usd_balance;
+    mapping(address => uint256) public crab_balance;
     struct Receipt {
-        address depositor;
+        address sender;
         uint256 amount;
     }
-    Receipt[] deposits;
-    Receipt[] crab_deposits;
+    Receipt[] public deposits;
```
Comment (KMKoushik) [inline packages/crab-netting/src/CrabNetting.sol:18]: `deposits.length` is not possible from UI. So we need the last index to loop and get receipts to show it to admin.

File: packages/crab-netting/src/CrabNetting.sol [hunk 13]
```diff
@@ -7,49 +9,126 @@ contract CrabNetting {
...
+    uint256 public depositsIndex;
+    Receipt[] public withdraws;
```
Comment (KMKoushik) [inline packages/crab-netting/src/CrabNetting.sol:20]: Similar to deposits, need last index for withdraws too.

File: packages/crab-netting/test/Netting.t.sol [hunk 14]
```diff
@@ -0,0 +1,121 @@
...
+
+        withdrawerPk = 0xA11CB;
+        withdrawer = vm.addr(withdrawerPk);
+
+        usdc.transfer(depositor, 400 * 1e18);
+        crab.transfer(withdrawer, 40 * 1e18);
+
+        vm.startPrank(depositor);
+        usdc.approve(address(netting), 200 * 1e18);
+        netting.depositUSDC(20 * 1e18);
```
Comment (KMKoushik) [inline packages/crab-netting/test/Netting.t.sol:46]: Need to change tests to 1e18

File: packages/crab-netting/test/Netting.t.sol [hunk 15]
```diff
@@ -0,0 +1,121 @@
...
+        netting.depositCrab(5 * 1e18);
+        netting.depositCrab(4 * 1e18);
+        netting.depositCrab(11 * 1e18);
+        assertEq(netting.crab_balance(withdrawer), 20e18);
+        vm.stopPrank();
+    }
+
+    function testNetting() public {
+        // TODO turn this into a fuzzing test
+        assertEq(usdc.balanceOf(withdrawer), 0, "starting balance");
```
Comment (KMKoushik) [inline packages/crab-netting/test/Netting.t.sol:63]: Should test partial withdraw flow too. 
- Deposit 100 USDC
- Withdraw 50 USDC
- Net the remaining amount

File: packages/crab-netting/src/CrabNetting.sol [hunk 16]
```diff
@@ -7,49 +9,126 @@ contract CrabNetting {
...
+        uint256 j = withdrawsIndex;
+        uint256 crab_quantity = forWithdraw / _price;
+        while (crab_quantity > 0 && j < withdraws.length) {
+            Receipt memory withdraw = withdraws[j];
+            if (withdraw.amount < crab_quantity) {
+                crab_quantity = crab_quantity - withdraw.amount;
+                crab_balance[withdraw.sender] =
+                    crab_balance[withdraw.sender] -
+                    withdraw.amount;
+                IERC20(usdc).transfer(
```
Comment (KMKoushik) [inline packages/crab-netting/src/CrabNetting.sol:95]: Need events to show PnL/ position to users. 

depositor, withdrawer, usdcAmount, crabAmount

File: packages/crab-netting/src/CrabNetting.sol [hunk 17]
```diff
@@ -7,49 +9,126 @@ contract CrabNetting {
     address usdc;
     address crab;
 
-    mapping(address => uint256) usd_balance;
-    mapping(address => uint256) crab_balance;
+    mapping(address => uint256) public usd_balance;
+    mapping(address => uint256) public crab_balance;
```
Comment (KMKoushik) [inline packages/crab-netting/src/CrabNetting.sol:13]: Not a big deal, we should be consistent with variable naming. Should be either camel case or snake case

**Review by KMKoushik — CHANGES_REQUESTED** [review]
Added some comments

File: packages/crab-netting/src/CrabNetting.sol (hunk 12 above)
Comment (sanandnarayan) [reply packages/crab-netting/src/CrabNetting.sol:18]: depositsQueued and withdraws Queued does this for you

File: packages/crab-netting/src/CrabNetting.sol (hunk 11 above)
Comment (sanandnarayan) [reply packages/crab-netting/src/CrabNetting.sol:19]: we are deleting an element, that element will be reset to its default value

File: packages/crab-netting/src/CrabNetting.sol (hunk 13 above)
Comment (sanandnarayan) [reply packages/crab-netting/src/CrabNetting.sol:20]: withdrawsQueued does this for you

**Pull request closed without merging**